cmd/collector/app/builder/builder_flags.go

```diff
@@ -24,18 +24,57 @@ import (
 	"flag"
 	"time"
 
+	"github.com/spf13/viper"
+
 	"github.com/uber/jaeger/cmd/collector/app"
 )
 
-var (
+const (
+	collectorQueueSize     = "collector.queue-size"
+	collectorNumWorkers    = "collector.num-workers"
+	collectorWriteCacheTTL = "collector.write-cache-ttl"
+	collectorPort          = "collector.port"
+	collectorHTTPPort      = "collector.http-port"
+)
+
+// CollectorOptions holds configuration for collector
+type CollectorOptions struct {
 	// QueueSize is the size of collector's queue
-	QueueSize = flag.Int("collector.queue-size", app.DefaultQueueSize, "The queue size of the collector")
+	QueueSize int
 	// NumWorkers is the number of internal workers in a collector
-	NumWorkers = flag.Int("collector.num-workers", app.DefaultNumWorkers, "The number of workers pulling items from the queue")
+	NumWorkers int
 	// WriteCacheTTL denotes how often to check and re-write a service or operation name
-	WriteCacheTTL = flag.Duration("collector.write-cache-ttl", time.Hour*12, "The duration to wait before rewriting an existing service or operation name")
+	WriteCacheTTL time.Duration
 	// CollectorPort is the port that the collector service listens in on for tchannel requests
-	CollectorPort = flag.Int("collector.port", 14267, "The tchannel port for the collector service")
+	CollectorPort int
 	// CollectorHTTPPort is the port that the collector service listens in on for http requests
-	CollectorHTTPPort = flag.Int("collector.http-port", 14268, "The http port for the collector service")
-)
+	CollectorHTTPPort int
+}
+
+// CreateFlags which are necessary CollectorOptions
+func CreateFlags() *flag.FlagSet {
+	flagSet := &flag.FlagSet{}
+	flagSet.Int(collectorQueueSize, app.DefaultQueueSize, "The queue size of the collector")
+	flagSet.Int(collectorNumWorkers, app.DefaultNumWorkers, "The number of workers pulling items from the queue")
+	flagSet.Duration(collectorWriteCacheTTL, time.Hour*12, "The duration to wait before rewriting an existing service or operation name")
+	flagSet.Int(collectorPort, 14267, "The tchannel port for the collector service")
+	flagSet.Int(collectorHTTPPort, 14268, "The http port for the collector service")
+	return flagSet
+}
+
+// NewCollectorOptions creates CollectorOptions with initialized properties
+func NewCollectorOptions(v *viper.Viper) *CollectorOptions {
+	// make sure that viper is initialized with default values for a case when CreateFlags is not
+	// passed to viper
```

Comment:
This is a little hack.

Normal configuration is:
1. create `cobra.Command`
2. add all necessary flags to the command
3. add command to viper
4. start command e.g. `cmd.Execute`
5. create configuration objects using viper

The problem is with default values because they are specified in flags. If we want to create configuration objects without creating flags beforehand then viper does have access to them. So I added this to ensure that default values are always passed to viper.